### PR TITLE
Watermarking changes for streaming

### PR DESCRIPTION
**Pull request by shirshanka**
Reworked watermark management to use fine-grained accounting to handle forks / converters correctly. 
Made AsyncWriterManager watermark aware. 
Added a Zk-based Watermark store using the existing ZkStateStore. 
Added sample pull files.

### CODE REVIEW

**Comment by shirshanka** [comment]
@abti @htran1 : please review when time permits.

File: gobblin-runtime/src/main/java/gobblin/runtime/StateStoreBasedWatermarkStorage.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    Config config = ConfigFactory.parseProperties(properties);
+
+    // Defaults
+    if (!config.hasPath(ConfigurationKeys.STATE_STORE_ROOT_DIR_KEY)) {
+      config = config.withValue(ConfigurationKeys.STATE_STORE_ROOT_DIR_KEY,
+          ConfigValueFactory.fromAnyRef("/streamingWatermarks"));
+    }
+    if (!config.hasPath("state.store.zk.connectString")) {
```
Comment (htran1) [inline gobblin-runtime/src/main/java/gobblin/runtime/StateStoreBasedWatermarkStorage.java]: Maybe remove this and add the default in ZkStateStoreFactory?

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java [hunk 2]
```diff
@@ -476,6 +487,13 @@ public void execute(JobExecutionContext context)
         throw new JobExecutionException(t);
       }
     }
+
+    @Override
+    public void interrupt()
+        throws UnableToInterruptJobException {
+      log.info("Job was interrupted");
```
Comment (htran1) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: Does this need to do anything other than the logging?

File: gobblin-runtime/src/main/java/gobblin/runtime/Task.java [hunk 3]
```diff
@@ -577,10 +628,13 @@ public String toString() {
...
   @SuppressWarnings("unchecked")
   private void processRecord(Object convertedRecord, ForkOperator forkOperator, RowLevelPolicyChecker rowChecker,
-      RowLevelPolicyCheckResults rowResults, int branches, CheckpointableWatermark watermark)
+      RowLevelPolicyCheckResults rowResults, int branches, AcknowledgableWatermark watermark)
       throws Exception {
     // Skip the record if quality checking fails
     if (!rowChecker.executePolicies(convertedRecord, rowResults)) {
+      if (watermark != null) {
+        watermark.ack();
+      }
```
Comment (abti) [inline gobblin-runtime/src/main/java/gobblin/runtime/Task.java]: not doing an ack on rowCheck = false will never ack a watermark

File: gobblin-cluster/src/main/java/gobblin/cluster/GobblinHelixJobLauncher.java [hunk 4]
```diff
@@ -138,6 +144,35 @@ public GobblinHelixJobLauncher(Properties jobProps, HelixManager helixManager, P
 
     this.taskStateCollectorService = new TaskStateCollectorService(jobProps, this.jobContext.getJobState(),
         this.eventBus, this.stateStores.taskStateStore, outputTaskStateDir);
+
+
+
+    // HACK to fixup IDEAL state with a rebalancer that will rebalance running jobs as well
+    final String clusterName = "GobblinStandaloneCluster";
```
Comment (shirshanka) [inline gobblin-cluster/src/main/java/gobblin/cluster/GobblinHelixJobLauncher.java]: done

File: gobblin-cluster/src/main/java/gobblin/cluster/GobblinHelixJobLauncher.java [hunk 5]
```diff
@@ -138,6 +144,35 @@ public GobblinHelixJobLauncher(Properties jobProps, HelixManager helixManager, P
...
+              idealState.setRebalancerClassName(rebalancerClassDesired);
+              helixAdmin.setResourceIdealState(clusterName, resource, idealState);
+            }
+
+             // HACK to ensure that concurrent tasks per instance is set.
+             // TODO: Remove this when we upgrade to a fixed Helix version
+             final String taskResourceName = TaskUtil.getNamespacedJobName(jobContext.getJobName(), jobContext.getJobId());
+             HelixConfigScope resourceScope = new HelixConfigScopeBuilder(HelixConfigScope.ConfigScopeProperty.RESOURCE)
+             .forCluster(clusterName).forResource(taskResourceName).build();
+             helixManager.getConfigAccessor().set(resourceScope, JobConfig.NUM_CONCURRENT_TASKS_PER_INSTANCE, "100");
```
Comment (shirshanka) [inline gobblin-cluster/src/main/java/gobblin/cluster/GobblinHelixJobLauncher.java]: added config.

File: gobblin-core-base/src/main/java/gobblin/writer/AcknowledgableWatermark.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import gobblin.source.extractor.CheckpointableWatermark;
+
+
+/**
+ * A Checkpointable Watermark that can be acknowledged. Useful for tracking watermark progress
+ */
+public class AcknowledgableWatermark implements Comparable<AcknowledgableWatermark>, Ackable {
+
+  private final CheckpointableWatermark _checkpointableWatermark;
+  private AtomicInteger _acked = new AtomicInteger(1);  // default number of acks needed is 1
```
Comment (shirshanka) [inline gobblin-core-base/src/main/java/gobblin/writer/AcknowledgableWatermark.java]: number of derived records per input record should not need a long.

File: gobblin-core-base/src/main/java/gobblin/writer/AcknowledgableWatermark.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    _acked.decrementAndGet();
+  }
+
+  public AcknowledgableWatermark incrementAck() {
+    _acked.incrementAndGet();
+    return this;
+  }
+
+  public boolean isAcked() {
+    return (_acked.get() == 0);
```
Comment (shirshanka) [inline gobblin-core-base/src/main/java/gobblin/writer/AcknowledgableWatermark.java]: count is 1 initially

File: gobblin-core-base/src/main/java/gobblin/writer/AcknowledgableWatermark.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  private final CheckpointableWatermark _checkpointableWatermark;
+  private AtomicInteger _acked = new AtomicInteger(1);  // default number of acks needed is 1
+
+  public AcknowledgableWatermark(CheckpointableWatermark watermark) {
+    _checkpointableWatermark = watermark;
+  }
+
+  @Override
+  public void ack() {
+    _acked.decrementAndGet();
```
Comment (shirshanka) [inline gobblin-core-base/src/main/java/gobblin/writer/AcknowledgableWatermark.java]: throwing assertion error now if this goes below 0 as it implies a bug in the code.

File: gobblin-core-base/src/main/java/gobblin/writer/MultiWriterWatermarkManager.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ * Responsible for managing continuous commit of watermarks.
+ * Periodically fetches watermarks from WatermarkAwareWriters and commits them to WatermarkStorage.
+ * TODO: Add metrics monitoring
+ */
+public class MultiWriterWatermarkManager implements WatermarkManager {
+
+
+  private final List<WatermarkAwareWriter> _watermarkAwareWriters;
```
Comment (shirshanka) [inline gobblin-core-base/src/main/java/gobblin/writer/MultiWriterWatermarkManager.java]: done.

File: gobblin-distribution/gobblin-flavor-custom.gradle [hunk 10]
```diff
@@ -33,8 +33,8 @@ dependencies {
 
   // Kafka integration (source, writer, metrics) -- choose one
   // compile project(':gobblin-modules:gobblin-kafka-08')
-  // compile project(':gobblin-modules:gobblin-kafka-09')
+  compile project(':gobblin-modules:gobblin-kafka-09')
```
Comment (shirshanka) [inline gobblin-distribution/gobblin-flavor-custom.gradle]: reverted

File: gobblin-runtime/src/main/java/gobblin/runtime/StateStoreBasedWatermarkStorage.java (hunk 1 above)
Comment (shirshanka) [inline gobblin-runtime/src/main/java/gobblin/runtime/StateStoreBasedWatermarkStorage.java]: done.

File: gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java (hunk 2 above)
Comment (shirshanka) [inline gobblin-runtime/src/main/java/gobblin/scheduler/JobScheduler.java]: for now just logging it.. will fix graceful shutdown in a separate pr.

File: gobblin-cluster/src/main/java/gobblin/cluster/GobblinHelixJobLauncher.java (hunk 4 above)
Comment (shirshanka) [inline gobblin-cluster/src/main/java/gobblin/cluster/GobblinHelixJobLauncher.java]: done

File: bin/gobblin-cluster-worker.sh [hunk 11]
```diff
@@ -70,19 +70,19 @@ do
       shift
```
Comment (shirshanka) [inline bin/gobblin-cluster-worker.sh]: done.

File: gobblin-runtime/src/main/java/gobblin/runtime/Task.java (hunk 3 above)
Comment (shirshanka) [inline gobblin-runtime/src/main/java/gobblin/runtime/Task.java]: not sure I understand... I am acking the watermark when the rowCheck returns false here. So I think I am handling this case.

**Pull request merged**